staging/src/k8s.io/apiserver/pkg/endpoints/filters/fail_with_audit.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/audit/policy"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// AuditWrapperFunc returns an http.Handler that audits a failed request.
+// The filters that appear before the audit filter in the handler chain can use it
+// to audit a failed request.
+// The specified errorHandler is invoked to render an error response to the request.
+// The specified stringer adapter provides an error message that is to be set in the
+// ResponseStatus field of the audit event.
+type AuditWrapperFunc func(errorHandler http.Handler, stringer func() string) http.Handler
+
+func (w AuditWrapperFunc) Wrap(handler http.Handler, stringer func() string) http.Handler {
+	return w(handler, stringer)
+}
+
+// FailWithAuditWrapper returns an AuditWrapperFunc that can be used to audit a failed request.
+func FailWithAuditWrapper(sink audit.Sink, checker policy.Checker) AuditWrapperFunc {
+	return failWithAuditWrapper(sink, checker, createAuditEventAndAttachToContext)
+}
+
+// auditEventCreateFunc defines the signature of the audit event create function used by the audit layer.
+// This function creates a new audit event from the given HTTP request and the policy checker.
+type auditEventCreateFunc func(*http.Request, policy.Checker) (*http.Request, *auditinternal.Event, []auditinternal.Stage, error)
+
+// audit create function is passed as a parameter, for better testability
+func failWithAuditWrapper(sink audit.Sink, checker policy.Checker, createFunc auditEventCreateFunc) AuditWrapperFunc {
+	return func(errorHandler http.Handler, stringer func() string) http.Handler {
+		if sink == nil || checker == nil {
+			return errorHandler
+		}
+
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
the above handler has been copied from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go#L35-L58 and made reusable, any server filter that appear before the audit filter can use it to audit a failed request.

If this looks good then I am happy to open a follow up PR to replace authn_audit.go.